sw/host/ot_certs/src/types.rs

```diff
@@ -0,0 +1,167 @@
+        VariableType::ByteArray { size: 0 }
+    }
+
+    fn random_value(size: usize) -> (Vec<u8>, Vec<u8>) {
+        // Generate a random byte array.
+        let bytes = (0..size).map(|_| rand::random::<u8>()).collect::<Vec<_>>();
+        let copy = bytes.clone();
+        (bytes, copy)
+    }
+}
+
+impl Type<BigUint> for BigUint {
+    fn variable_type() -> VariableType {
+        // ASN1 integers are always in big-endian.
+        VariableType::Integer { size: 0 }
+    }
+
+    fn random_value(size: usize) -> (Vec<u8>, BigUint) {
+        // Generate a random big integer but make sure that the MSB is nonzero
+        // so that the ASN1 encoder will not shrink its size.
+        let mut val = (1..size).map(|_| rand::random::<u8>()).collect::<Vec<_>>();
+        val.insert(0, 0xff);
+        let big_int = BigUint::from_bytes_be(&val);
+        (val, big_int)
+    }
+}
+
+impl Type<String> for String {
+    fn variable_type() -> VariableType {
+        VariableType::String { size: 0 }
+    }
+
+    fn random_value(size: usize) -> (Vec<u8>, String) {
+        let bytes = (0..size)
+            .map(|_| rand::distributions::Alphanumeric.sample(&mut rand::thread_rng()))
+            .collect::<Vec<_>>();
+        let string = std::str::from_utf8(&bytes)
+            .expect("Alphanumeric returned invalid ASCII characters!")
+            .to_string();
+        (bytes, string)
+    }
+}
+
+fn check_conversion(
+    source: &VariableType,
+    target: VariableType,
+    convert: &Option<Conversion>,
+    allowed_convert: &Vec<Option<Conversion>>,
+) -> Result<VariableType> {
+    if !allowed_convert.contains(convert) {
+        bail!(
+            "cannot convert from {:?} to {:?} using conversion {:?}, possible conversions are {:?}",
+            source,
+            target,
+            convert,
+            allowed_convert
+        );
+    }
+    Ok(target)
+}
+
+// Given a source variable type, an optional conversion and a target variable type
+// (of unspecified size), return either an error if this conversion does not make
+// sense, or return the target variable type with the size that corresponds to the
+// input size after conversion.
+pub fn convert_type(
+    source: &VariableType,
+    target: &VariableType,
+    convert: &Option<Conversion>,
+) -> Result<VariableType> {
+    let no_conv = vec![None];
+    let endian_conv = vec![Some(Conversion::LittleEndian), Some(Conversion::BigEndian)];
+    match target {
+        VariableType::ByteArray { .. } => {
+            // A conversion to a byte array is always possible and does not change
+            // the size. No conversion operator allowed.
+            check_conversion(
+                source,
+                VariableType::ByteArray {
+                    size: source.size(),
+                },
+                convert,
+                &no_conv,
+            )
+        }
+        VariableType::String { .. } => {
+            match source {
```

Comment:
Do `match *source` here to save some ugly `size: *size` later.

Reply:
With the rewrite you suggested, I now have some matches on `VariableType::String { size }` where I don't know if this trick works?

Reply:
You can also match `&VariableType::String { size }` which gives you a `size` binding without reference.